Network file browser

hectorcorrea opened the pull request:
This is now ready to be merged.

This PR implements a "tree-view" for the Network File Brower so that the user can select individual files inside folders. When the page is loaded only the files and folders on the user's home directory are shown (as it used to work before). The additional functionality kicks in when user click on a folder icon, at that point we load the files and folders (via AJAX) for that particular folder. The user then can select individual files/folders to upload.

Below is an screenshot of how it looks: 

![screen shot 2015-07-08 at 4 55 39 pm](https://cloud.githubusercontent.com/assets/568286/8582035/4fb7b086-2592-11e5-9777-a020b31d5a4f.png)

mjgiarlo commented:
Impressive!  I wonder if this feature might be extractable to Sufia at some point -- I could see folks being widely interested in this.

awead commented on app/views/generic_files/upload/_local_file_import_chooser.html.erb:
```diff
@@ -0,0 +1,259 @@
+<% end %>
+
+<script type="text/javascript">
+
```
put the js in the asset pipeline.

awead commented on app/views/generic_files/upload/_local_file_import_chooser.html.erb:
```diff
@@ -0,0 +1,259 @@
+<style type="text/css">
+  input, select, textarea, th, td {
```
css should go in the asset pipeline.

awead commented on app/controllers/local_files_controller.rb:
```diff
@@ -0,0 +1,33 @@
+
+  def load_path
+    path = params[:path]  
+    if valid_path?(path)
```
maybe use `@path` instead and validate it when you set it. Then your logic can just be `if @path` or `unless @path`

awead commented:
@hectorcorrea this looks great! I'd say it's pretty close to done. I just made a few comments.

hectorcorrea commented:
@mjgiarlo @awead This version is almost there but **don't merge it just yet** 

I've polished the code a bit and addressed your comments. Notice that I purposefully left the styles on the the page since (1) they are not reused and (2) I didn't want to accidentally alter the style of other parts of the code. I'll look at this with @mtribone next week.

Most of the JavaScript was moved to its own file so that the asset pipeline can load it. I left the "document on ready" part inside the Ruby file because I don't want that code to be executed on every page. We should talk about this next week. If I understand our code correctly, the rest of our pages don't follow this approach which means we are executing unnecessarily document on ready left and right for pages that don't need it. 

I used what I think is a more appropriate structure in which most of the JavaScript code leaves on an external files (handled by the asset pipeline) except the "document on ready" part that kicks the JavaScript initialization.

awead commented on app/views/generic_files/upload/_local_file_import_chooser.html.erb:
```diff
@@ -0,0 +1,116 @@
+  # on this file (rather than under assets) because it
+  # is tighly coupled to the display of the tree used
+  # on this page and we are not reusing them.
+  #
```
It's still better to have all the css in the pipeline. If you're concerned that it'll overwrite existing styles, using Sass would ensure that it only applies to these selectors within specific classes.

The other reason for the pipeline is that all the css will only be loaded once by the browser. When include styles in the view, they get loaded and parsed every time the page loads and can increase load times.

hectorcorrea replied:
Yeah, I might submit a separate PR later on to move this to the asset pipeline. For now I am thinking of leaving them here so that this PR can be merged. It's getting to be too big as it is and it provides already a better UI than what we had. I've documented this as a separate issue https://github.com/psu-stewardship/archivesphere/issues/50

awead commented on app/views/generic_files/upload/_local_file_import_chooser.html.erb:
```diff
@@ -0,0 +1,116 @@
+
+<script type="text/javascript">
+  $(document).ready(function() {
+    // Javascript array with initial set of files/folders to render.
```
You can make this a JQuery plugin and include it into generic_files.js, similar to:

https://github.com/projecthydra/sufia/blob/master/app/assets/javascripts/sufia/uploader.js#L24

Then, load the function based on a selector in the page, like:

https://github.com/projecthydra/sufia/blob/master/app/assets/javascripts/sufia/fileupload.js

That way the js only gets loaded when the page requires it.

hectorcorrea replied:
Those are good options. We should definitively consider this when/if we move this to a generic module to allow others to re-use it.

awead replied:
It's not for re-use, but for the same reasons I outlined above about loading the script each time the page loads. You also use `<%= javascript_tag do %>` for the script block.

hectorcorrea replied:
I've updated the code to use the Rails `javascript_tag` helper as you suggested. 

I am not sure the helper adds any clarity in this particular example (if anything it prevents SublimeText from getting the syntax highlight right) but so be it. If the team prefers the Rails helpers to raw HTML/JavaScript I am here to please :)

hectorcorrea commented:
OK. I've posted what I think is a first good pass to be merged and shown to Ben for feedback. I've added a couple of extra tickets (#50 #51) to enhance the functionality but I think this is good to go.
